### ISSUE TITLE
Add static types

### ISSUE THREAD

**Issue by mpeyper** [ioof-holdings/redux-subspace#15]
After attending a [talk at the YOW! conference](https://a.confui.com/-j0wpqARu) I was compelled to try out [flow](https://flowtype.org/) and [typescript](https://www.typescriptlang.org/) to add more confidence for me when writing javascript code (I come from a mostly Java/C# background).

I would like some comment and review of:
[flow](https://github.com/mpeyper/redux-subspace/tree/static-types/flow)
[typescript](https://github.com/mpeyper/redux-subspace/tree/static-types/typescript)

I'm particularly interested in your thoughts on which you prefer and why.  When I get time, I'll add my thoughts.

Is this something we actually want for this library? I do, but this was more of a thought experiment for me personally and I'm sure there are improvements that can be made to my first attempts.

Just adding the types identified a few issues in the existing code, which I'll issues for (probably with pull requests - they're pretty minor) so I definitely see a benefit of it.

**Comment by vivian-farrell** [comment]
Personally I'm not a fan of typescript mainly because I think the extra code and overhead outweighs the benefits. It's not used much in the react community but is required for angular 2. Not sure if there would be much benefit for this library unless typescript was also used on the consuming end as well. Maybe exposing the ts files would be a nice extra but unless people really need it, it could just be more overhead.

**Comment by mpeyper** [comment]
That's interesting to me because I would say that the extra code (of both Typescript and Flow - there isn't much difference in my opinion) add a huge amount of benefit to me in the form of confidence that the code does what I think it does.  I would even say that if you are using ES6, the amount of extra code is fairly minimal.

I the argument of Typescript vs. Flow, I definitely agree that for a library to be consumed by other things (React vs. Angular 2 aside), Typescript might not be a sensible choice.  I actually thought afterwards that it might be more prudent to keep the tests in plain javascript as that's where the majority of the bugs would occur (in the consumer's javascript).

Flow on the other hand slots into the babel lifecycle a bit better (read: not at all for Typescript).  I did find Typescript's type inference was better than Flow's and I found myself having to work around a few of the quirks.  For example, [here(https://github.com/mpeyper/redux-subspace/blob/static-types/flow/src/utils/dispatch.js#L51) flow did not think `action` was a callable function when it was called [here](https://github.com/mpeyper/redux-subspace/blob/static-types/flow/src/utils/dispatch.js#L54) until i specifically typed it to a `Thunk`.  

It also isn't as strict as Typescript and allows some things through.  An example of this is providing additional parameters to function calls.  Typescript makes you remove them unless you provide specific overrides that handle them.  [Flow allows them by default](https://flowtype.org/docs/functions.html#too-many-arguments) which leads to confusing calls and a messier codebase.

Having said all that, I feel like Flow is a good balance for the people that prefer strictly typed systems (like myself) and those that don't (like @vivian-farrell, I assume).
